File: driver-api/src/main/scala/net/cakesolutions/reactiveredis/driver/api/results.scala
package net.cakesolutions.reactiveredis.driver.api

import scala.collection.JavaConverters._

object results {
  sealed trait RedisResult
  case class Result[T](r: T) extends RedisResult
  case class CombinedResult(results: List[RedisResult]) extends RedisResult

  sealed trait StatusResult extends RedisResult
  case object OK extends StatusResult
  case object NOK extends StatusResult

  def result[T](r: T) = Result(r)
  def combinedResult(results: List[RedisResult]) = CombinedResult(results)
  def jcombinedResult(results: java.util.List[RedisResult]) = combinedResult(results.asScala.toList)
}

File: driver-api/src/main/java/net/cakesolutions/reactiveredis/driver/japi/Results.java
package net.cakesolutions.reactiveredis.driver.japi;

import net.cakesolutions.reactiveredis.driver.api.results;
import net.cakesolutions.reactiveredis.driver.api.results.RedisResult;
import scala.Option;

import java.util.List;
import java.util.Optional;

public class Results {
    public static RedisResult ok() {
        return results.OK$.MODULE$;
    }

    public static RedisResult nok() {
        return results.NOK$.MODULE$;
    }

    public static <T> RedisResult result(T r) {
        return results.result(r);
    }

    public static RedisResult combinedResult(List<RedisResult> redisResults) {
        return results.jcombinedResult(redisResults);
    }

    public static <T> Optional<T> toJava(Option<T> option) {
        return option.isEmpty() ? Optional.<T>empty() : Optional.of(option.get());
    }
}
